### PR TITLE
ILMarshaler Refactor

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Refactor the interop `ILMarshaler`s to have a more consistent structure. Add some comments for how the marshaler system is structured.

This could be a first step towards unifying our parameter/return value marshalling with our field marshalling. Or at least towards making the code easier to reason about for any future path we take with the marshalers.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
@dotnet-bot help

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

File: src/vm/ilmarshalers.h [hunk 1]
```diff
@@ -1417,7 +1474,16 @@ class ILMarshaler
...
+    virtual bool AlwaysConvertByValContentsCLRToNative()
+    {
+        LIMITED_METHOD_CONTRACT;
+        return false;
+    }
+
 public:
+    
+    // Extension point to allow a marshaler to conditionally override all of the ILMarshaler logic with its own or block marshalling when marshalling an argument.
+    // See mlinfo.cpp:3523 for the implementation.
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.h]: Like the comment - remove the line number and replace with function name.

File: src/vm/ilmarshalers.h [hunk 2]
```diff
@@ -1781,73 +1847,73 @@ class ILReflectionObjectMarshaler : public ILMarshaler
...
+    BinderClassID GetManagedTypeBinderID() override { LIMITED_METHOD_CONTRACT; return CLASS__FIELD_HANDLE; }
 };
 
 class ILBoolMarshaler : public ILMarshaler
 {
 public:
 
     virtual CorElementType GetNativeBoolElementType() = 0;
-    virtual int GetNativeTrueValue() = 0;
+    virtual int GetNativeTrueValue() = 0 ;
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.h]: Rogue space?

File: src/vm/ilmarshalers.h (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/vm/ilmarshalers.h]: I think there are a few other spots this happened - i.e. line number.

Comment (jkoritzinsky) [reply src/vm/ilmarshalers.h]: I only remember doing that on this line and again a few lines lower. I'll go verify now.

Comment (jkoritzinsky) [reply src/vm/ilmarshalers.h]: Yeah those were the only two cases of that in this PR.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test this please

**Comment by jkoritzinsky** [comment]
Removed some changes from this PR that aren't as useful and rebased on master. 

Diff is significantly easier to read with whitespace changes ignored (approximately half the size of what it was before the rebase).

File: src/vm/ilmarshalers.h [hunk 3]
```diff
@@ -2832,26 +2939,42 @@ class ILAsAnyMarshalerBase : public ILMarshaler
...
+    bool NeedsClearNative() override final;
+    void EmitClearNativeTemp(ILCodeStream* pslILEmit) override final;
+    bool AlwaysConvertByValContentsCLRToNative() override final
+    {
+        LIMITED_METHOD_CONTRACT;
+        return true;
+    }
+
+private:
+    DWORD GetAsAnyFlags()
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.h]: can be marked `const`?

File: src/vm/ilmarshalers.h [hunk 4]
```diff
@@ -2832,26 +2939,42 @@ class ILAsAnyMarshalerBase : public ILMarshaler
...
     };
 
-    ILAsAnyMarshalerBase() :
-        m_dwMarshalerLocalNum(-1)
-    {
-        LIMITED_METHOD_CONTRACT;
-    }
-
 protected:
     static const BYTE ML_IN  = 0x10;
```
Comment (AaronRobinsonMSFT) [inline src/vm/ilmarshalers.h]: Can you move both `ML_IN` and `ML_OUT` to usage and document why the values are what they are.

**Comment by jkoritzinsky** [comment]
The CI failures are a known issue. I'll retrigger CI once they're fixed and run some tests on some WinForms and WPF apps to validate behavior.

**Comment by jkoritzinsky** [comment]
I've run this branch against the WinForms integration test suite and against one apps in the WPF AppCompat suite and everything works as expected. If CI is green after I address your comments, I'll merge this in.

**Comment by AaronRobinsonMSFT** [comment]
@stephentoub This PR inserted the behavioral change you are now observing regarding excessive allocations.

@jkoritzinsky This has severely impacted a scenario involving `HttpClient`. We need to either fix the regression or back this out ASAP. Based on @stephentoub's allocation profile the underlying issue appears to be excessive allocations in the code below:

https://github.com/dotnet/coreclr/blob/f163abbd16d35032e4450a09c548cb804cf36a85/src/System.Private.CoreLib/src/System/StubHelpers.cs#L1558-L1563

**Comment by stephentoub** [comment]
I'm seeing these getting allocated now for every input SafeHandle marshaled, e.g. for every call we make on a socket.  Is that expected?  We need to not do that ;)
cc: @jkotas

**Pull request merged**